libs/blocks/merch/merch.js

```diff
@@ -37,16 +37,33 @@ export async function loadEntitlements() {
     fetch(entitlementsURL),
   ]).then(([{ default: getUserEntitlements }, mappings]) => {
     if (!mappings.ok) return [];
-    return Promise.all([getUserEntitlements(), mappings.json()]);
+    return Promise.all([getUserEntitlements({ params: [{ name: 'include', value: 'OFFER.PRODUCT_ARRANGEMENT' }], format: 'raw' }), mappings.json()]);
   });
   return loadEntitlements.promise;
 }
 
 async function getCheckoutAction(offers) {
-  const [{ arrangement_codes: aCodes } = {}, entitlementsMappings] = await loadEntitlements();
-  if (aCodes === undefined) return undefined;
+  let upgradeAction;
+  const [entitlements, entitlementsMappings] = await loadEntitlements();
+  const aCodes = entitlements?.map((offer) => offer.offer.product_arrangement_code);
   const [{ productArrangementCode }] = offers;
-  if (aCodes[productArrangementCode] === true) {
+  const [{ productArrangement }] = offers;
+  const ctaProductFamily = productArrangement?.productFamily;
+  const upgradeOffer = await document.querySelector('.merch-offers.upgrade [data-wcs-osi]')?.onceSettled()
+    .catch((e) => {
+      window.lana.log('Failed to resolve an upgrade offer:', e);
+      return undefined;
+    });
+  if (upgradeOffer && Array.isArray(entitlements) && ctaProductFamily) {
+    const { default: handleUpgradeOffer } = await import('./upgrade.js');
+    upgradeAction = handleUpgradeOffer(ctaProductFamily, upgradeOffer, entitlements);
+  }
+  if (upgradeAction) return upgradeAction;
```

Comment:
whole reason why this test can be false and this strange 57-61 block logic is here, is because handleUpgradeOffer could get from buildUrl that both upgrade offer & origin offer have an id. Can't we make this cleaner / simpler by extracting the source & target offer extraction from entitlement logic and line 57 testing if there are any, then line 59 we know we have all that we need and we can return the url.
In other words, could handleUpgradeOffer just keep the offer extraction logic, and return target offer (or undefined),
and could we return buildURL l59? 

wdyt?

Reply:
sorry, i didn't understand your suggestion.
do you mean to move the code below into handleUpgradeOffer in `upgrade.js`?
```
const upgradeOffer = await document.querySelector('.merch-offers.upgrade [data-wcs-osi]')?.onceSettled()
    .catch((e) => {
      window.lana.log('Failed to resolve an upgrade offer:', e);
      return undefined;
    });
```

>In other words, could handleUpgradeOffer just keep the offer extraction logic, and return target offer (or undefined),

handleUpgradeOffer returns an action object which contains cta text, and a function that renders the modal. In your suggestion if only URL is returned by handleUpgradeOffer, where will the modal be created?

Reply:
as discussed offline, I could move the checks for eligibility to merch.js and have them all run before we load upgrade.js, but then part of upgrade logic will be in one file and another part in another file.
which is ok for me, although I lean more towards keeping as much as possible in upgrade.js
would like to know also what others think.